Trie/Trie/Trie/Trie.swift

```diff
@@ -0,0 +1,207 @@
+//
+//  Trie.swift
+//  Trie
+//
+//  Created by Rick Zaccone on 2016-12-12.
+//  Copyright © 2016 Rick Zaccone. All rights reserved.
+//
+
+import Foundation
+
+
+/// A node in the trie
+class TrieNode<T: Hashable> {
+    var value: T?
+    weak var parentNode: TrieNode?
+    var children: [T: TrieNode] = [:]
+    var isTerminating = false
+    var isLeaf: Bool {
+        get {
+            return children.count == 0
+        }
+    }
+
+    /// Initializes a node.
+    ///
+    /// - Parameters:
+    ///   - value: The value that goes into the node
+    ///   - parentNode: A reference to this node's parent
+    init(value: T? = nil, parentNode: TrieNode? = nil) {
+        self.value = value
+        self.parentNode = parentNode
+    }
+
+    /// Adds a child node to self.  If the child is already present,
+    /// do nothing.
+    ///
+    /// - Parameter value: The item to be added to this node.
+    func add(value: T) {
+        guard children[value] == nil else {
+            return
+        }
+        children[value] = TrieNode(value: value, parentNode: self)
+    }
+}
+
+/// A trie data structure containing words.  Each node is a single
+/// character of a word.
+class Trie {
+    typealias Node = TrieNode<Character>
+    /// The number of words in the trie
+    public var count: Int {
+        return wordCount
+    }
+    /// Is the trie empty?
+    public var isEmpty: Bool {
+        return wordCount == 0
+    }
+    /// All words currently in the trie
+    public var words: [String] {
+        return wordsInSubtrie(rootNode: root, partialWord: "")
+    }
+    fileprivate let root: Node
+    fileprivate var wordCount: Int
+
+    /// Creats an empty trie.
+    init() {
+        root = Node()
+        wordCount = 0
+    }
+}
+
+// MARK: - Adds methods: insert, remove, contains
+extension Trie {
+
+    /// Inserts a word into the trie.  If the word is already present,
+    /// there is no change.
+    ///
+    /// - Parameter word: the word to be inserted.
+    func insert(word: String) {
+        guard !word.isEmpty else {
+            return
+        }
+        var currentNode = root
+        let charactersInWord = Array(word.lowercased().characters)
+        var index = 0
+        while index < charactersInWord.count {
```

Comment:
A `for` loop can do exactly what we're trying to do in a cleaner manner in this scenario (I know the site tutorial uses the `while` construct, but a fellow contributor alerted me of a better way).

```
for character in word.lowercased().characters {
  if let childNode = currentNode.children[character] {
    currentNode = childNode
  } else {
    currentNode.add(value: character)
    currentNode = currentNode.children[character]!
  }
}
```

This removes the necessity of the `index` and `charactersInWord` properties. The `for` construct also helps us bind each element in the `word.lowercased().characters`, alleviating the necessity of picking it out ourselves.